scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,353 @@
+        + 0.1 * (x[:, 2]**4) * np.sin(x[:, 0])
+    )
+    return f_eval.reshape(-1, 1)
+
+
+def sample_A_B(
+    d: int, n: int,
+    dists: Optional[List[PINVDist]] = None,
+    random_state: SeedType = None
+) -> Tuple[np.ndarray,  np.ndarray]:
+    """Sample two matrices A and B.
+
+    Uses a Sobol' sequence with 2`d` columns to have 2 uncorrelated matrices.
+    This is more efficient than using 2 random draw of Sobol'.
+    See sec. 5 from [1]_.
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    A_B = qmc.Sobol(d=2*d, seed=random_state, bits=64).random(n)
+
+    A, B = A_B[:, :d], A_B[:, d:]
+
+    if dists is not None:
+        for d_, dist in enumerate(dists):
+            dist_rng = NumericalInversePolynomial(
+                dist, random_state=random_state
+            )
+            A[:, d_] = dist_rng.ppf(A[:, d_])
+            B[:, d_] = dist_rng.ppf(B[:, d_])
+
+    return A, B
+
+
+def sample_AB(A: np.ndarray, B: np.ndarray) -> np.ndarray:
+    """AB matrix.
+
+    AB: columns of B into A. Shape (n*d, d). e.g in 2d
+    Take A and replace 1st column with 1st column of B. You have (n, d)
+    Then A and replace 2nd column with 2nd column of B. You have (n, d)
+    Concatenate to get AB. Which means AB is (2*n, d).
+    """
+    n, d = A.shape
+    AB = np.empty((int(d*n), d))
+    for i in range(d):
+        AB[i*n:(i+1)*n, :] = np.column_stack((A[:, :i], B[:, i], A[:, i+1:]))
+
+    return AB
+
+
+def saltelli_2010(
+    f_A: np.ndarray, f_B: np.ndarray, f_AB: np.ndarray
+) -> Tuple[np.ndarray,  np.ndarray]:
+    """Saltelli2010 formulation.
+
+    .. math::
+
+        S_i = \frac{1}{N} \sum_{j=1}^N
+        f(\mathbf{B})_j (f(\mathbf{AB}^{(i)})_j - f(\mathbf{A})_j)
+
+    .. math::
+
+        S_{T_i} = \frac{1}{N} \sum_{j=1}^N
+        (f(\mathbf{A})_j - f(\mathbf{AB}^{(i)})_j)^2
+
+    References
+    ----------
+    .. [1] Saltelli, A., P. Annoni, I. Azzini, F. Campolongo, M. Ratto, and
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, f_A.shape[0])
+
+    var = np.var(np.vstack([f_A, f_B]))
+
+    s = np.mean(f_B * (f_AB - f_A.flatten()).T, axis=0) / var
+    st = 0.5 * np.mean((f_A.flatten() - f_AB).T ** 2, axis=0) / var
+
+    return s, st
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    d: IntNumber,
+    dists: Optional[List[PINVDist]] = None,
+    l_bounds: Optional[npt.ArrayLike] = None,
+    u_bounds: Optional[npt.ArrayLike] = None,
```

Comment:
I haven't looked closely yet at what is going on, but I'm guessing that all the information in `d` is contained in `l_bounds` and `u_bounds` if they are provided (i.e. `len(l_bounds)`). If both are provided, they would need to be checked for consistency, right? And the default bounds would be `(0, 1)` if `None` are provided - how often do you think that default would be useful?

Instead, if you're willing to require that bounds be provided explicitly, `d`, `l_bound`, and `u_bounds` could be replaced by a single required `bounds` parameter which accepts instances of `scipy.optimize.Bounds`.

Reply:
Yes bounds, `d` and `dists` are linked. That could be a good compromise to use `Bounds` 👍 

Ideally, a single params for `dists` would be better. This would be closer to how this is used in practice.

Most people use a default uniform distribution and don't have scaled parameters (this does not make a difference). What does make a difference is the type of distribution. But most of the time, either people don't know about stats and distribution or they just have no clue about the distribution of their data (seen it too much in real life sadly...). The worst part is that it has an impact on the result OC (I should add a warning there to encourage people to set this).

Anyway, if we would force people to give `dists`, we can avoid `d` at all. Or we only force to provide `bounds` and by default it uses uniform dists. That would also work. Maybe better.

Reply:
Ok. Maybe if you document the current parameters it would be easier to decide.

Reply:
I updated the doc, hopefully this is helpful. I will also add some tests to show how it can be used.